Revert to C++ 17

Koranir opened the pull request:
**CI/CD/Testing**

This PR addresses the bug/feature described by @thewierdnut on discord

## Summary
Reverts C++ 20 relevant changes.

Zireael07 commented:
Can you explain for those of us who are not on Discord?

RisingLeaf commented on source/Engine.cpp:
```diff
@@ -100,10 +100,36 @@ namespace {
-	constexpr auto PrunePointers = [](auto &objects) { erase_if(objects,
-			[](const auto &obj) { return obj->ShouldBeRemoved(); }); };
-	constexpr auto Prune = [](auto &objects) { erase_if(objects,
-			[](const auto &obj) { return obj.ShouldBeRemoved(); }); };
```
i think it might make sense to keep these in a comment for the day when we actually do the switch(when enough support), cause I really like them.

Koranir commented:
> Can you explain for those of us who are not on Discord?

Right, Endless Sky's mobile port, made by thewierdnut wants to keep supporting android version below 5 (which a fair amount of people use, apparently). The benefits brought by C++ 20 aren't big enough that dropping support for these people is worth it (it's mostly just convenience functions that are easily recreatable).

warp-core commented:
I'm not inclined to make a change that is probably going to lead to many people having problems compiling the game (I expect this to happen because it happened when we added CMake, it happened when we switched to C++17, and it happened when we switched to C++20) so soon before a planned release when supporting such users would likely take time away from fixing any actual bugs that show up. Particularly when the reason for the change is to enable a fork to more easily maintain compatibility with very old devices.

If the changes here are correct, it doesn't appear that the current codebase has diverged too far from C++17; if supporting those old devices is important right now, I expect the endless-mobile release for 0.10.9 can use the commits here.

Koranir commented:
> I'm not inclined to make a change that is probably going to lead to many people having problems compiling the game (I expect this to happen because it happened when we added CMake, it happened when we switched to C++17, and it happened when we switched to C++20) so soon before a planned release when supporting such users would likely take time away from fixing any actual bugs that show up. Particularly when the reason for the change is to enable a fork to more easily maintain compatibility with very old devices.
> 
> If the changes here are correct, it doesn't appear that the current codebase has diverged too far from C++17; if supporting those old devices is important right now, I expect the endless-mobile release for 0.10.9 can use the commits here.

How would this give users trouble when compiling the game? At most it would be a cache clean, it's not like C++ 20 capable compilers suddenly make errors with C++ 17. And it's because we haven't diverged so far that I think it's alright to just go back to C++ 17 (we won't be missing out on anything too major, unlike C++ 14->17 when we got std::filesystem, etc).

thewierdnut commented:
> Can you explain for those of us who are not on Discord?

Specifically, in order to get a compiler that supports c++20, endless-mobile would have to fast-forward to android-ndk 26, which drops support for kit-kat.

Koranir commented:
> Is it really worth supporting a 10 year old obsolete version?
> 
> https://composables.com/android-distribution-chart says kitkat is 0.30%
> 
> Wikipedia claims "As of October 2022, 1.39% of Android devices run KitKat." so two years later I bet the percentage is even lower now

We kinda target the lowest common denominator,  and it doesn't really hurt anybody if we revert to C++ 17, so even if 1% more people can use ES it's a perfectly reasonable compromise.

thewierdnut commented:
Since we released without this merge request, I would just as soon drop it, as I'm going to have to push a new version with the updated ndk anyways. Thanks for trying though.

tibetiroka commented:
Closing as it's no longer relevant.
